csrc/scheduler/matmul_heuristic.h

```diff
@@ -0,0 +1,160 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <c10/util/hash.h>
+#include <mma_type.h>
+#include <scheduler/heuristic.h>
+#include <utils.h>
+#include <functional>
+
+#include <sstream>
+#include "type.h"
+
+namespace nvfuser {
+
+// Parameters of the matmul heuristic to describe the optimial schedule.
+class MatmulParams : public HeuristicParams {
+ public:
+  //! A list of possible strategies used to define along which axis
+  //!  parallelization will be done.
+  enum class TileRasterizationOrder { RowMajor = 0, ColumnMajor = 1 };
+
+  //! A wrapper for double buffering config pieces
+  struct DoubleBufferOptions {
+    bool double_buffer_smem_write = false;
+    bool double_buffer_smem_read = false;
+    int smem_double_buffer_stage = 2;
+
+    bool operator==(const DoubleBufferOptions& other) const {
+      return other.double_buffer_smem_write == double_buffer_smem_write &&
+          other.double_buffer_smem_read == double_buffer_smem_read &&
+          other.smem_double_buffer_stage == smem_double_buffer_stage;
+    }
+
+    std::string toString() const {
+      std::stringstream ss;
+      ss << "DoubleBufferOptions:\n"
+         << "  double_buffer_smem_write: "
+         << (double_buffer_smem_write ? "true" : "false") << "\n"
+         << "  double_buffer_smem_read: "
+         << (double_buffer_smem_read ? "true" : "false") << "\n"
+         << "  smem_double_buffer_stage: " << smem_double_buffer_stage;
+      return ss.str();
+    }
+
+    size_t hash() const {
```

Comment:
Very naive, but why not going with `std::hash<>` specializations instead ?

Reply:
I agree. We should change that (but not in this PR), not only for matmul, but for all schedulers.

Reply:
Such specialization will require extending namespace `std` that I'm a little bit hesitant to do (although this is the way we are supposed to do it, if I'm not wrong). Maybe it's not a bad idea and we could have all matmul related structs, that require hashing, have own specialization - this can be checked and done as a follow up of this PR. Does it sound ok?